### ISSUE TITLE
Add revokable api token to accounts for headless servers

### ISSUE THREAD

**Issue by gentlecolts** [Yellow-Dog-Man/Resonite-Issues#1284]
### Is your feature request related to a problem? Please describe.

Currently headless servers require username/password stored in plaintext via the config file.  This is insecure, and puts headless accounts at risk in the event of a breach

### Describe the solution you'd like

Accounts should instead be able to generate secure tokens, which the headless server could then use as a means of authentication.  Auth via token would then be able to restrict functionality, preventing attackers from taking full control over the account and allowing the owner to revoke the compromised token as needed

### Describe alternatives you've considered

#253 brings up 2fa as another security enhancement, however i don't see this as an acceptable alternative, as a breach of the server would still leave passwords exposed

### Additional Context

zangooseoo

**Comment by Frooxius** [comment]
It's something that I've been considering generally for a bit, though I'm not sure which way would be best to approach this from user perspective.

One approach would be that you log in once manually with the headless and the headless will then just save the token and keep re-using that. But that requires manual intervention every time you need to start fresh.

The other would be that you do request a login token for your account. This one could be permanent or expire at some point. And you just copy it there. However it will only prevent certain things - e.g. you won't be able to change password on the account or request certain things if the user is logged in with that token, but you can still access quite a bit of account.

You'll be able to revoke that token.

So one question is, how will the process for generating the token actually work? We could just do a command in the Resonite bot for managing them for now.

The other part is restricting functionality - what are you actually looking to restrict? Because headless needs most account functionality right now. We'd need to formalize on what can be restricted and how.

**Comment by gentlecolts** [comment]
A headless should only need to be able to load assets and interact with contacts, I may be missing some other actions but it shouldn't need to do any "privledged" actions like updating passwords/contact info/2fa or creating new api tokens

Consider now that users can use their main accounts as headless hosts as well, the headless shouldn't be considered to be a true "user", but an application that is using the user's account.  This may also go into the new enhanced settings ui, and we could do the following:

- On the settings page, include an api tokens section, possible categories to place this page could include developer or security
- On this page, a user authenticated via password would be able to generate api tokens and give them a friendly name, and view existing applications
- When generating a new token, they should be presented with the ability to give the token a friendly name, and the ability to copy the generated token for use in the config file
- The user should not be able to ever view the token again after initial creation, tokens should only be stored on Resonite's servers and never be retrievable through any api or logging.  The list displaying active tokens should ONLY display the friendly name, and present the option to revoke the token

**Comment by gentlecolts** [comment]
It may also be a good idea that, if a headless is using username/password in the config, that there should be some mechanism by which it can use this full auth to automate this token generation, store in the config, and remove the more sensitive user password

**Comment by Frooxius** [comment]
Settings is one way to do it yes, but that will have to wait for the new one. We plan to have a section for configuring "login methods", so that could fit in there.

And yes, updating passwords is definitely something that shouldn't be allowed, as well as new tokens. 

However headless might need to do stuff like update contacts - e.g. when you add some people so they can join and so on.

**Comment by gentlecolts** [comment]
> However headless might need to do stuff like update contacts

Correct, but like, for example, would a headless ever need to add or remove items from an authenticated user's inventory?  Interact with currency-related APIs if those are ever added?  I'd guess not.

Read/Write access to contact apis and read-only access to asset ones seems like the proper scope to me

**Comment by Nytra** [comment]
Headless does need to be able to save worlds too
